minimal_data_sharing/minimal_atomic_parameter.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Carlos San Vicente
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+static_assert(std::atomic<std::uint32_t>::is_always_lock_free);
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count_++);
+        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
+        this->publisher_->publish(message);
+      };
+
+    realtime_callback_group_ = this->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive, false);
+    timer_ = this->create_wall_timer(500ms, timer_callback, realtime_callback_group_);
+
+    count_ = declare_parameter("count", 0);
+    parameter_event_handler_ = std::make_shared<rclcpp::ParameterEventHandler>(this);
+    auto param_callback = [this](const rclcpp::Parameter & p) {
+        RCLCPP_INFO(this->get_logger(), "Count changed to: '%ld'", p.as_int());
+        count_.store(static_cast<std::uint32_t>(p.as_int()));
```

Comment:
When is this callback called? In case of the concurrent modification the last updated value wins but anyone concurrently reading (publishing) may see it (eventually) or see a later modification.

What I do not understand is that apparently this callback modifies `count_`, but also publish itself (operator++). Is this intended? What is the use case? I would imagine the parameter changes independently of publishing and the publish call itself does not change it.

Reply:
>When is this callback called? In case of the concurrent modification the last updated value wins but anyone concurrently reading (publishing) may see it (eventually) or see a later modification.

The parameter callback is invoked when the parameter is set externally (the parameter is a service under the hood). For example:

```
ros2 param set /minimal_publisher count 0
```

>What I do not understand is that apparently this callback modifies count_, but also publish itself (operator++). Is this intended? What is the use case? I would imagine the parameter changes independently of publishing and the publish call itself does not change it.

In this example, I wanted to show a ROS parameter that is used from a real-time thread (publishing) but also can be modified from a non-real-time thread (parameter callback).

I used the [minimal_publisher](https://github.com/ros2/examples/blob/master/rclcpp/topics/minimal_publisher/lambda.cpp) as a base to construct this example, so the count was the first thing that came to my mind to make it a parameter. It is probably not a meanful example so we can change this with other thing so it's more descriptive. As we do with the PID parameter example.

One option could be to use something like a "multiplier" parameter. The count would be a normal value which is increased in each publishing. We publish `mulitplier x count`. The multiplier value could be changed from the parameter callback.

What do you think?

Reply:
@shuhaowu @MatthiasKillat I updated this example. I re-used the example from 
- https://hogliux.github.io/farbot/presentations/meetingcpp_2019/assets/player/KeynoteDHTMLPlayer.html#54
- https://github.com/hogliux/farbot#realtimeobject

In this case, there is nothing interesting to show in the example output besides some sine wave data. We could simply show some code snippets.